File: demos/react-native/modules/offline/ios/OfflineCore/StreamSource.swift
import Foundation

public final class StreamSource: ByteSource {
    public private(set) var fileSize: Int64 = 0
    public private(set) var contentType: String = "audio/mp4"

    private let session: StreamSession
    private var bootstrapped = false

    public init(workerUrl: URL,
                tokenProvider: @escaping @Sendable () async throws -> String,
                trackRef: StreamSession.TrackRef,
                urlSession: URLSession = .shared) {
        self.session = StreamSession(
            workerUrl: workerUrl, tokenProvider: tokenProvider,
            trackRef: trackRef, urlSession: urlSession)
    }

    /// Forces bootstrap; populates fileSize + contentType. Idempotent.
    public func prepare() async throws {
        if bootstrapped { return }
        let boot = try await session.bootstrap()
        fileSize = boot.fileSize
        contentType = boot.contentType
        bootstrapped = true
    }

    public func read(range: Range<Int64>) async throws -> Data {
        if !bootstrapped { try await prepare() }
        return try await session.read(range: range)
    }

    public func close() {
        let s = session
        Task { await s.close() }
    }

    /// Exposed so the player can drive heartbeats on a timer.
    public func heartbeat() async throws -> Bool { try await session.heartbeat() }

    public func sessionMetrics() async -> StreamSession.SessionMetrics {
        await session.metrics()
    }
}

File: modules/offline-core-android/src/main/kotlin/cc/musicme/offline/StreamSource.kt
package cc.musicme.offline

import okhttp3.OkHttpClient

class StreamSource(
    workerUrl: String,
    tokenProvider: suspend () -> String,
    trackRef: TrackRef,
    client: OkHttpClient = OkHttpClient(),
) : ByteSource {

    private val session = StreamSession(workerUrl, tokenProvider, trackRef, client)
    private var bootstrapped = false
    override var fileSize: Long = 0
        private set
    override var contentType: String = "audio/mp4"
        private set

    suspend fun prepare() {
        if (bootstrapped) return
        val b = session.bootstrap()
        fileSize = b.fileSize
        contentType = b.contentType
        bootstrapped = true
    }

    override suspend fun read(start: Long, endExclusive: Long): ByteArray {
        if (!bootstrapped) prepare()
        return session.read(start, endExclusive)
    }

    suspend fun heartbeat(): Boolean = session.heartbeat()
    fun metrics(): SessionMetrics = session.metrics()
    override fun close() { session.close() }
}
